openvasd/openvasd.c

```diff
@@ -0,0 +1,2251 @@
+          curl_easy_cleanup (curl);
+          *err = g_strdup (
+            "{\"error\": \"Not possible to set Client certificate\"}");
+          return NULL;
+        }
+      blob.data = (*conn)->key;
+      blob.len = strlen ((*conn)->key);
+      blob.flags = CURL_BLOB_COPY;
+
+      if (curl_easy_setopt (curl, CURLOPT_SSLKEY_BLOB, &blob) != CURLE_OK)
+        {
+          g_warning ("%s: Not possible to set the Client private key",
+                     __func__);
+          curl_easy_cleanup (curl);
+          *err = g_strdup (
+            "{\"error\": \"Not possible to set Client private key\"}");
+          return NULL;
+        }
+    }
+
+  // Set API KEY
+  if ((*conn)->apikey)
+    {
+      xapikey = g_string_new ("X-API-KEY: ");
+      g_string_append (xapikey, (*conn)->apikey);
+      customheader = curl_slist_append (customheader, g_strdup (xapikey->str));
+      g_string_free (xapikey, TRUE);
+    }
+
+  switch (method)
+    {
+    case POST:
+      if (data != NULL && data[0] != '\0')
+        {
+          // SET Content type
+          customheader =
+            curl_slist_append (customheader, "Content-Type: application/json");
+          // Set body
+          curl_easy_setopt (curl, CURLOPT_POSTFIELDS, data);
+          curl_easy_setopt (curl, CURLOPT_POSTFIELDSIZE, strlen (data));
+        }
+      break;
+    case GET:
+      curl_easy_setopt (curl, CURLOPT_HTTPGET, 1L);
+      break;
+    case DELETE:
+      curl_easy_setopt (curl, CURLOPT_CUSTOMREQUEST, "DELETE");
+      break;
+    default:
+      curl_easy_setopt (curl, CURLOPT_CUSTOMREQUEST, "HEAD");
+      break;
+    };
+
+  if (customheader != NULL)
+    curl_easy_setopt (curl, CURLOPT_HTTPHEADER, customheader);
+
+  // Init the struct where the response is stored and set the callback function
+  curl_easy_setopt (curl, CURLOPT_WRITEFUNCTION, response_callback_fn);
+  curl_easy_setopt (curl, CURLOPT_WRITEDATA, resp);
+
+  return curl;
+}
+
+/** @brief Send request
+ *
+ *  @param[in] curl        The CURL handler to perform an request.
+ *  @param[in] header_name If this field is set, is looked in the header and
+ *                         its value is returned inside the response.
+ *  @param[out] response   The response struct to be filled with the response
+ * code and the header value.
+ *
+ *  @return Return struct containing the http response code and the response
+ * body. In case of error the struct is filled with code RESP_CODE_ERR (-1) and
+ * a message. NULL on memory related error. Response must be free()'ed by the
+ * caller with openvasd_response_free()
+ */
+static openvasd_resp_t
+openvasd_send_request (CURL *curl, const char *header_name,
+                       openvasd_resp_t response)
+{
+  long http_code = RESP_CODE_ERR;
+
+  int ret = CURLE_OK;
+  if ((ret = curl_easy_perform (curl)) != CURLE_OK)
+    {
+      g_warning ("%s: Error sending request: %d", __func__, ret);
+      curl_easy_cleanup (curl);
+      curl = NULL;
```

Comment:
```suggestion
```
Nitpick: This makes me think you're trying to do something to the structure pointed at by curl.